xds/internal/client/xds.go

```diff
@@ -452,7 +458,12 @@ func routesProtoToSlice(routes []*v3routepb.Route, logger *grpclog.PrefixLogger,
 			case *v3routepb.HeaderMatcher_ExactMatch:
 				header.ExactMatch = &ht.ExactMatch
 			case *v3routepb.HeaderMatcher_SafeRegexMatch:
-				header.RegexMatch = &ht.SafeRegexMatch.Regex
+				regex := ht.SafeRegexMatch.GetRegex()
```

Comment:
I was thinking we can be a bit more aggressive here.

Options:
1. (probably not a good idea): we turn the match fields into a [`matcher`](https://github.com/grpc/grpc-go/blob/d426aa5f2e5e809639b45d9619416ce22e56319a/xds/internal/resolver/matcher.go#L33), and put that in the RDS update struct, so the resolver can use directly.
   - this is bad because the log probably won't be as easy to read, and also it may be hard to compare RDS updates in the tests
1. (better and easier than 1?): change the type of `header.RegexMatch` to `regexp.Regexp`
   - this saves the `regexp.Compile()` in resolver (and the error handling)
   - this has the similar problem for tests. But I think we can just compare the result of `String()`

Reply:
Done.

Reply:
Hmm, I'm surprised that we don't need to fix tests for this (the newer commit). Maybe we don't have Regex success tests, and the two Regex failure tests aren't affected. Can you double check?

Reply:
Added a test for the regex success case.